source/common/upstream/upstream_impl.cc

```diff
@@ -576,6 +576,23 @@ void PrioritySetImpl::updateHosts(uint32_t priority, UpdateHostsParams&& update_
   static_cast<HostSetImpl*>(host_sets_[priority].get())
       ->updateHosts(std::move(update_hosts_params), std::move(locality_weights), hosts_added,
                     hosts_removed, overprovisioning_factor);
+
+  if (!batch_update_) {
+    runUpdateCallbacks(hosts_added, hosts_removed);
+  }
+}
+
+void PrioritySetImpl::updateHostsNoCallbacks(uint32_t priority,
```

Comment:
Just as a reference to the reviewer:
PrioritySetImpl::updateHosts() may run same callback twice if batch_update_ is false, not sure about the use case for that.
We could try to define a BatchUpdateHelper, like eds.cc, but that seems complicated.
It might be better if we can have an interface to set batch_update_ to true.

Reply:
The reason this was happening was because I was building hosts_added incorrectly, as you pointed out below. This is fixed, so I reverted these changes. Resolving.